### DISCUSSION TITLE
Shutting down UV when using UDP handles

### DISCUSSION THREAD

**Discussion by sbonar** [libuv/libuv#4071] [Q&A] [answered]
I have come across an issue that I hope someone can help me with.  
I am more than likely not doing something correctly when shutdown down UV when using UDP handles.
I call uv_stop()
then I loop thru the list of UDP handles I believe are currently open.
I call uv_udp_recv_stop()
if the handle is active I then call uv_close() on it.  
*NOTE I added the "is active" check because I found that if the handle is not active on a close it will assert.  Is this correct?
Then I call uv_loop_close() and check for UV_EBUSY.  If it is busy I use uv_walk() to loop thru the remaining handles.
In the walk CB I again needed to check if the handle was active or not before calling close

Even after doing all that I still am getting back an UV_EBUSY from uv_loop_close().  

What am I doing wrong?

**Comment by sbonar** [comment]
After some more debugging I figured out that the uv_close() was asserting, NOT because the handle was still active, BUT because it was already in a closing state.
I guess then the next question is what would prevent the UDP handle from closing?

**Comment by sbonar** [comment]
It almost looks like the handle is NOT ACTIVE so it's not going to finish closing. 
My current implementation is NOT queuing my sends but instead just calling uv_udp_try_send().  In this case should I not be calling uv_udp_recv_stop() before calling uv_close()?

**Comment by sbonar** [comment]
Unfortunately removing the calls to uv_udp_recv_stop() did not fix this issue.

**Comment by sbonar** [comment]
I guess I should also preference this I am running uv_run() within a thread and will continue to call uv_run() until that thread is shutdown.  It's this thread shutdown calls the uv_stop() and closes (or attempts to close) the handles.

As an example:
while(thread_is_running)
{
    sleep_for_1ms;
    uv_run(UV_RUN_DEFAULT);
}

When I stop the thread and try to join it, it at this point that everything is hanging because uv_run() is never exiting to see that "thread_is_running" is false to get out of the while loop

**Reply by bnoordhuis** [reply]
Are you calling libuv methods from other threads than the one that calls uv_run()?

**Reply by sbonar** [reply]
Yes, one, uv_udp_try_send().  Is this not allowed? I guess a better question would be is running uv_run in a thread like this incorrect?

**Reply by sbonar** [reply]
Well thinking about it some more, I do make other UV API calls from other threads.  How much is libuv not thread safe?

**Reply by bnoordhuis** [reply, marked as answer]
Only uv_async_send() is thread-safe, nothing else that operates on an event loop is.

**Comment by sbonar** [comment]
Thanks.  Back to the drawing board. :-). 
All part of the learning process.
